pkg/translator/loki/logs_to_loki.go

```diff
@@ -22,13 +22,141 @@ import (
+// and "loki.resource.labels". Each hint might contain a comma-separated list of
+// attributes (resource or record) that should be promoted to a Loki label. Those
+// attributes are removed from the body as a result, otherwise they would be shown
+// in duplicity in Loki.
+// PushStreams are created based on the labels: all records containing the same
+// set of labels are part of the same stream. All streams are then packed within
+// the resulting PushRequest.
+// When this function isn't able to marshal a log record, the log record is dropped
+// and processing continues, so that the caller can decide to either skip the entire
+// batch or send only the data that could be parsed. The caller can use the PushReport
+// to make this decision, as it includes all of the errors that were encountered,
+// as well as the number of items dropped and submitted.
+func LogsToLokiRequests(ld plog.Logs) map[string]PushRequest {
+	groups := map[string]pushRequestGroup{}
+
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		ills := rls.At(i).ScopeLogs()
+
+		for j := 0; j < ills.Len(); j++ {
+			logs := ills.At(j).LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+
+				// similarly, we may remove attributes, so change only our version
+				log := plog.NewLogRecord()
+				logs.At(k).CopyTo(log)
+
+				// we may remove attributes, so we make a copy and change our version
+				resource := pcommon.NewResource()
+				rls.At(i).Resource().CopyTo(resource)
+
+				// resolve tenant and get/create a push request group
+				tenant := getTenantFromTenantHint(log.Attributes(), resource.Attributes())
+				group, ok := groups[tenant]
+				if !ok {
+					group = pushRequestGroup{
+						report:  &PushReport{},
+						streams: make(map[string]*logproto.Stream),
+					}
+					groups[tenant] = group
+				}
+
+				mergedLabels := convertAttributesAndMerge(log.Attributes(), resource.Attributes())
+				// remove the attributes that were promoted to labels
+				removeAttributes(log.Attributes(), mergedLabels)
+				removeAttributes(resource.Attributes(), mergedLabels)
+
+				// create the stream name based on the labels
+				labels := mergedLabels.String()
+
+				entry, err := convertLogToJSONEntry(log, resource)
+				if err != nil {
+					// Couldn't convert so dropping log.
+					group.report.Errors = append(group.report.Errors, fmt.Errorf("failed to convert, dropping log: %w", err))
+					group.report.NumDropped++
+					continue
+				}
+
+				group.report.NumSubmitted++
+
+				if stream, ok := group.streams[labels]; ok {
+					stream.Entries = append(stream.Entries, *entry)
+					continue
+				}
+
+				group.streams[labels] = &logproto.Stream{
+					Labels:  labels,
+					Entries: []logproto.Entry{*entry},
+				}
+			}
+		}
+	}
+
+	requests := make(map[string]PushRequest)
+	for tenant, g := range groups {
+		pr := &logproto.PushRequest{
+			Streams: make([]logproto.Stream, len(g.streams)),
+		}
+
+		i := 0
+		for _, stream := range g.streams {
```

Comment:
why index from the range is not used here? `for i, stream := range g.streams {`

Reply:
`streams` is a map, so `i` is a key that we ignore